admin/app/dfp/CustomTargetingAgent.scala

```diff
@@ -0,0 +1,69 @@
+package dfp
+
+import com.google.api.ads.dfp.axis.utils.v201705.StatementBuilder
+import com.google.api.ads.dfp.axis.v201705.{CustomTargetingKey, CustomTargetingValue}
+import common.Logging
+import common.dfp.{GuCustomTargeting, GuCustomTargetingValue}
+
+import scala.util.Try
+
+object CustomTargetingAgent extends DataAgent[Long, GuCustomTargeting] with Logging {
+
+  def loadFreshData(): Try[Map[Long, GuCustomTargeting]] = Try {
+    val maybeData = for (session <- SessionWrapper()) yield {
+
+      val keys: Map[Long, CustomTargetingKey] = session.customTargetingKeys(new StatementBuilder()).map{ key => key.getId.longValue -> key}.toMap
+
+      val statementWithIds = new StatementBuilder()
+        .where(s"customTargetingKeyId IN (${keys.keys.mkString(",")})")
+
+      val valuesByKey: Map[Long, Seq[CustomTargetingValue]] = session.customTargetingValues(statementWithIds).groupBy{_.getCustomTargetingKeyId.longValue}
+
+      valuesByKey flatMap { case (keyId: Long, values: Seq[CustomTargetingValue]) =>
+        keys.get(keyId) map { key =>
+
+          val guValues: Seq[GuCustomTargetingValue] = values map {
+            value => GuCustomTargetingValue(
+              id = value.getId.longValue,
+              name = value.getName,
+              displayName = value.getDisplayName)
+          }
+
+          keyId -> GuCustomTargeting(
+              keyId = keyId,
+              name = key.getName,
+              displayName = key.getDisplayName,
+              values = guValues)
+        }
+      }
+    }
+
+    maybeData getOrElse Map.empty
+  }
+}
+
+object CustomTargetingService {
+
+  def targetingKey(session: SessionWrapper)(keyId: Long): String = {
+    lazy val fallback = {
```

Comment:
Yeah, the call to `session.customTargetingKeys` is actually executing a call to the DFP API, so we only want to call this when we haven't got the data in the agent.